### PR TITLE
Refactor for scraped

### PR DESCRIPTION
**Pull request by ondenman**
Part of: https://github.com/everypolitician/everypolitician-data/issues/25298

### CODE REVIEW

File: lib/member_name_parts.rb [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  def female_prefixes
+    %w(Mrs Ms Miss)
+  end
+
+  def male_prefixes
+    %w(Mr)
+  end
+
+  def gender_map
+    Hash[female_prefixes.map { |e| [e, 'female'] }].merge(Hash[male_prefixes.map { |e| [e, 'male'] }])
```
Comment (tmtmtmtm) [inline lib/member_name_parts.rb]: As this is now using a more modern Ruby, you should be able to simplify these with `.to_h`

File: lib/member_page.rb [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+  field :area_id do
+    region_id_and_region.first
+  end
+
+  field :area do
+    region_id_and_region.last
+  end
+
+  field :term do
+    11
```
Comment (tmtmtmtm) [inline lib/member_page.rb]: This should either be extracted from the page, or removed from the Page

File: lib/member_page.rb [hunk 3]
```diff
@@ -0,0 +1,90 @@
+# frozen_string_literal: true
+require 'scraped'
+
+class MemberPage < Scraped::HTML
+  # decorator Scraped::Response::Decorator::AbsoluteUrls
```
Comment (tmtmtmtm) [inline lib/member_page.rb]: if this isn't needed, there's no real value in leaving it here commented out.

File: lib/members_page.rb [hunk 4]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+require 'scraped'
+
+class MembersPage < Scraped::HTML
+  decorator Scraped::Response::Decorator::AbsoluteUrls
+
+  field :member_urls do
+    noko.css('.who-parliament .member-image .swap-title a/@href').map(&:text).each do |link|
+      link
+    end
```
Comment (tmtmtmtm) [inline lib/members_page.rb]: What's this `.each` loop doing?

File: scraper.rb [hunk 5]
```diff
@@ -6,118 +6,25 @@
...
+  klass.new(response: Scraped::Request.new(url: url).response)
 end
 
 def scrape_list(url)
-  noko = noko_for(url)
-  noko.css('.who-parliament .member-image .swap-title a/@href').map(&:text).each do |link|
-    scrape_mp(URI.join(url, link))
+  (scrape url => MembersPage).member_urls.each do |mem_url|
+    data = (scrape mem_url => MemberPage).to_h
+    warn [data[:honorific_prefix], data[:name], data[:gender]].join(' ----- ')
```
Comment (tmtmtmtm) [inline scraper.rb]: Probably best to remove this, or replace it with the (commented out) version that actually lets us check for consistency across all the data:
 
```ruby
puts data.reject { |k, v| v.to_s.empty? }.sort_by { |k, v| k }.to_h
```

File: lib/member_name_parts.rb [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    prefixes_chomped.map { |p| gender_map[p] }.compact.first
+  end
+
+  private
+
+  def prefixes
+    %w(Assoc Prof Professor Rev Bishop Prince Dr Lt Col Colonel (Ret’)).to_set
+                                                                       .merge(female_prefixes)
+                                                                       .merge(male_prefixes)
+  end
```
Comment (tmtmtmtm) [inline lib/member_name_parts.rb]: This is operating at several different levels of specificity at the same time. It would be better to split this so there are three different sets of prefixes which then get joined in one place.

**Review by tmtmtmtm — CHANGES_REQUESTED** [review]

File: lib/member_name_parts.rb [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+require 'scraped'
+
+class MemberNameParts < Scraped::HTML
+  field :prefix do
+    partitioned.first
+  end
+
+  field :name do
+    # .sub called for specific case where Dr. is not separated from name by space
+    partitioned.last.sub('Dr.Bharrat', 'Bharrat')
```
Comment (tmtmtmtm) [inline lib/member_name_parts.rb]: This doesn't really seem like the best place to fix that sort of thing up.

**Review by tmtmtmtm — CHANGES_REQUESTED** [review]
This whole thing is massively complex. Most of this is because the original version was also needlessly complex, but this is pushing it in the wrong direction I fear. 

A lot of this is simply because I wrote the original when I didn't know the ruby standard library very well, so I think I'd simplify it all to something more like:

```ruby
class MemberNameParts < Scraped::HTML
  field :prefix do
    partitioned.first.join(' ')
  end

  field :name do
    partitioned.last.join(' ')
  end

  field :gender do
    return 'male' if (prefixes & MALE_PREFIXES).any?
    return 'female' if (prefixes & FEMALE_PREFIXES).any?
  end

  private

  OCCUPATIONAL_PREFIXES = %w(Assoc Prof Professor Rev Bishop Prince Dr Lt Col Colonel (Ret’))
  FEMALE_PREFIXES       = %w(Mrs Ms Miss)
  MALE_PREFIXES         = %w(Mr)
  PREFIXES              = OCCUPATIONAL_PREFIXES + FEMALE_PREFIXES + MALE_PREFIXES

  def partitioned
    words.partition { |w| PREFIXES.include? w.chomp('.') }
  end

  def prefixes
    partitioned.first.map { |w| w.chomp('.') }
  end

  def words
    noko.text.tidy.sub(/^Hon.? /, '').split(/\s/)
  end
end
```

(Though I suspect some of it could still be simplified further.)

I think we definitely need some tests for this, though.

File: scraper.rb [hunk 8]
```diff
@@ -6,118 +6,28 @@
...
 
 def scrape_list(url)
-  noko = noko_for(url)
-  noko.css('.who-parliament .member-image .swap-title a/@href').map(&:text).each do |link|
-    scrape_mp(URI.join(url, link))
+  (scrape url => MembersPage).member_urls.each do |mem_url|
+    data = (scrape mem_url => MemberPage).to_h
+                                         .merge(term: 11)
+    # .sub! called for specific case where Dr. is not separated from name by space
+    data[:name].sub!('Dr.Bharrat', 'Bharrat')
```
Comment (tmtmtmtm) [inline scraper.rb:26]: This isn't just an EP-world fixup: anyone using the classes should get the correct name/prefix for this person.

NB: as this is an edge case you're going to need a test for it.

File: scraper.rb [hunk 9]
```diff
@@ -6,118 +6,28 @@
...
+    puts data.reject { |_k, v| v.to_s.empty? }.sort_by { |k, _v| k }.to_h
```
Comment (tmtmtmtm) [inline scraper.rb:27]: This should be commented out for production.

**Review by tmtmtmtm — CHANGES_REQUESTED** [review]

**Comment by ondenman** [comment]
Closing in favour of: https://github.com/everypolitician-scrapers/guyana-parliament/pull/4
